### PR TITLE
Warning when using automatic template polymorphism

### PR DESCRIPTION
**Pull request by SkySkimmer**
I also annotated every spot in the stdlib/test suite where it triggered. Probably most of them don't need it. Should I try doing some filtering?
I also disabled template polymorphism for types declared by funind, this seems to work fine.

I'm not sure that we actually want this. Thoughts? What's the next step for template polymorphism?
Should we try disabling the automatic decision by default (and make PRs to 3rd parties to add the option or annotate with attributes)?

### CODE REVIEW

**Comment by SkySkimmer** [comment]
>I'm not sure that we actually want this.

Essentially, I do think we want to move away from silent template polymorphism but when I made this PR it seemed somehow heavy (and I abandoned annotating the non-output test suite files, which trigger the warning about 300-400 times).

**Review by ppedrot — APPROVED** [review]
I like the concept, and the code looks OK.

**Review by gares — COMMENTED** [review]
I like the idea.
I think the name of the attribute could be better.
Either `universes-template-polymorphic` or `universes(template)` if `universes(polymorphic)` enables the other kind of polymorphism.
Just `template` is too ambiguous (template coq?).
Finally, I still hope template polymorphism goes away, hence `universes(template)` seems a better option.

**Comment by gares** [comment]
FTR, the `polymorphic` attribute is also badly named, since polymorphism means another thing to most of us, so maybe this is the right occasion to introduce an attribute that describes how *universes* are interpreted.

**Comment by SkySkimmer** [comment]
Now using `#[universes(template)]`

**Comment by herbelin** [comment]
> Now using #[universes(template)]

In passing, don't you want to have a true discussion on the syntax of attributes? No need to move Coq to jargon-like syntaxes. At least asking users if they really prefer this kind of syntax over, say, a minima, `#[Template Polymophic]`?

**Comment by maximedenes** [comment]
> In passing, don't you want to have a true discussion on the syntax of attributes? No need to move Coq to jargon-like syntaxes. At least asking users if they really prefer this kind of syntax over, say, a minima, `#[Template Polymophic]`?

Of course, we can discuss. I personally find "template polymorphism" more jargon than parentheses...

**Comment by Zimmi48** [comment]
I think it is important that the discussion happens, otherwise it will always feel like it forced its way through. Here is not the place to have the discussion of course. We should probably switch back to #8027.

**Comment by maximedenes** [comment]
> I think it is important that the discussion happens

Sure, I really wasn't trying to suggest otherwise. My point was simply that for the scope of this PR, the terminology and concepts we use (template polymorphism) seems way more intimidating to me than the syntax.

**Comment by herbelin** [comment]
My remark was on the attribute syntax. On the PR, rephrasing @skyskimmer himself, I'm not sure that we actually want this.

Even if we want to forbid `Prop` <= `Set`, I don't understand well why (the accordingly intimidating) so-called template polymorphism could not be transparently simulated from universe polymorphism and cumulativity?

**Comment by SkySkimmer** [comment]
>I don't understand well why (the accordingly intimidating) so-called template polymorphism could not be transparently simulated from universe polymorphism and cumulativity?

The behaviour wrt module subtyping is especially different. I'm sure there are other differences.

**Comment by herbelin** [comment]
> My point was simply that for the scope of this PR, the terminology and concepts we use (template polymorphism) seems way more intimidating to me than the syntax.

I hope that you will agree that it may be worth to be honest and not to mix the two questions though...

**Comment by mattam82** [comment]
@herbelin @SkySkimmer on the topic of template-polymorphism versus poly+cumulativity, yes we should be able now to emulate it. I think we should rather try to turn template-polymorphic inductives into polymorphic cumulative ones for a start and see where this goes. This is not my top priority right now but anyone is welcome to try. If that goes well, we can think of deprecating. What is the issue with module subtyping you are refering to @SkySkimmer ?

**Comment by SkySkimmer** [comment]
Subtyping universe polymorphic stuff is generally a pain as instances need to match, this simply does not happen with template polymorphism.

**Comment by mattam82** [comment]
But it's rare enough to use a non-trivial subtyping of inductives no? Does it happen in the stdlib for example? Note that I'm proposing to make only the Inductives polymorphic, not the associated definitions yet.

**Comment by mattam82** [comment]
@herbelin it might be necessary, however the internal (global) universe will have to be fixed anyway in both cases so I'm not sure it changes much.

**Comment by herbelin** [comment]
> the internal (global) universe will have to be fixed anyway 

I'm sorry, I don't get what you mean by "fixed"?

**Comment by mattam82** [comment]
I mean in unifications of T@{outer inner} = T@{outer' inner'}, the inner universes will have to be equal, while in the template polymorphic mode you'd have a single representative of the inner universe.

**Comment by JasonGross** [comment]
> I think we should rather try to turn template-polymorphic inductives into polymorphic cumulative ones for a start and see where this goes

I frequently take advantage of the fact that `sig` lands in `Prop` when both of it's arguments do, but unlike `ex` can be eliminated when the goal lives in `Type`.  This would be lost with universe polymorphism, right?

**Comment by mattam82** [comment]
@JasonGross that's right indeed, we'd need to allow Prop to instantiate universe binders for this. Why don't you use a proper sigma type in Prop instead?

**Comment by JasonGross** [comment]
> Why don't you use a proper sigma type in Prop instead?

Mostly convenience.  Because the standard library doesn't provide one (let alone have nice notation for it), and I have a decent amount of automation infrastructure around the two existing sigma types.

**Comment by ppedrot** [comment]
It looks like the comments of this PR have been hijacked to a mildly unrelated topic, which is emulation of template polymorphism using cumulativity. Notwithstanding this problem, what is the future of this PR? Do we still want the explicit attribute or what?

**Comment by mattam82** [comment]
Sorry for the hijack. In the current situation, template is not subsumed by cumulative polymorphism, so I'm not sure we want to activate the warning by default for users just yet but certainly the functionality is useful.

**Comment by SkySkimmer** [comment]
I think we do want to move towards making use of template polymorphism explicit, even if we're not ready to deprecate template polymorphism itself.

**Comment by mattam82** [comment]
Hmm, shouldn't there be a notemplate attribute as well here? Users should be able to avoid it, at least to prevent breaking extraction.

**Comment by SkySkimmer** [comment]
notemplate already exists.

**Comment by mattam82** [comment]
Ok, my bad! I guess I'd like to have opinions of users here, given that the warning will impact every definition of an inductive in Type in the wild, I guess that will come already from users of master when me merge. One last thing, should the warning be disabled in compat/Coq89.v?

**Comment by SkySkimmer** [comment]
The warning is meant to be used to transition away from automatic 
template polymorphism, so after putting #[template] on the ones where 
it's wanted they can turn off the option.

I don't know if we have any policy on warnings wrt compat settings.

Gaëtan Gilbert

**Comment by mattam82** [comment]
@herbelin Currently we cannot implement template using cumulative polymorphism, so this is only about making explicit when template polymorphism is used. The idea is that users should now be able to specify how their inductive is interpreted w.r.t. universes, and that information appears in the source. I think making this explicit is a plus. With respect to future plans, we'd like template to be replaced with cumulative polymorphic inductive types in the future, and in that case, it shall only be a matter of moving from universes(template) to universes(polymorphic, cumulative). I don't think the default should become polymorphic, cumIulative though, as that would change the meaning of existing definitions which use global universes. Similarly, if we decide to implement your suggestion to emulate template with a restricted kind of polymorphic, cumulative (being polymorphic only on the universes in the parameters that appear in the conclusion) then we'll still want to distinguish those two modes.

In short, this is about being explicit rather than implicit. And quoting you, I feel the consensus is that no form of polymorphism should be the default indeed, to avoid surprises. The proposed warning acts as a kind of `Set Template Polymorphic Inductives` option, when explicitly disabled.

**Comment by herbelin** [comment]
@mattam82: Thanks for clarifying your plans, that seems good.

**Comment by SkySkimmer** [comment]
So is this getting merged?

**Comment by ppedrot** [comment]
I guess this is essentially waiting for a green light from @herbelin.

**Comment by herbelin** [comment]
As far as the objective announced by @mattam82 in his comment above is to have no form of polymorphism at all on by default, a warning (and the PR) is ok to me.

**Comment by ppedrot** [comment]
@SkySkimmer Can you rebase to run the CI on a fresh master?

**Comment by ppedrot** [comment]
AAC-tactics is broken by this PR, this needs an overlay.

**Comment by SkySkimmer** [comment]
Backward compatible overlay waiting for merge at https://github.com/coq-community/aac-tactics/pull/27

**Comment by ppedrot** [comment]
@SkySkimmer can you rebase to get a fresh CI run?

**Comment by SkySkimmer** [comment]
Rebase again and again...

**Review by gares — APPROVED** [review]

**Comment by Zimmi48** [comment]
I too have realized after the fact that I had a few problems with this PR. The first one is that things are seriously badly documented. The second one is that it should be possible to use the option to say I want template polymorphism everywhere and I don't want the warning (currently the solution is to silence the warning). I've tried to come up with a solution with a non-boolean option: people could do `Set Auto Template Polymorphism "enabled"`, `Set Auto Template Polymorphism "disabled"`, or `Unset Auto Template Polymorphism` (default, which emits a warning if there is no attribute). However, I've hit strange failures...

**Comment by herbelin** [comment]
Coming back to this PR, wouldn't there be a way to make the warning more precise? For instance, I get a warning `Automatically declaring True as template polymorphic` if I do
```coq
Inductive True := I.
```
But the warning seems to come here only from putting `True` in `Prop`, not from using any kind of polymorphism at all.

As far as I understand the objectives, it shall be either (1) to ask users to write explicitly
```coq
Inductive True : Prop := I.
```
to have `True` in `Prop`, or (2) to write:
```coq
Inductive unit : Type := tt.
```
or
```coq
Inductive unit : Set := tt.
```
to have `unit` not formally put in `Prop`.

I guess the warning could be something like `Automatically declaring True in Prop but the default shall change. Use explicitly Prop or Set to avoid the warning.`. Then, if the objective is (1) to have the warning whenever `Type` or no explicit type is given, and, if the objective is (2), to have it displayed only when `Type` is given.

What do you think?

**Comment by SkySkimmer** [comment]
>Coming back to this PR, wouldn't there be a way to make the warning more precise?

Indeed this is a tricky case as thinking about it led me to #9294 so I'm not sure if we can detect it. Maybe if we do like with the "could not define as primitive record" warning where we look at what the kernel inferred.
We would also want to avoid warning when there is literally no difference between template and monomorphic, eg `Inductive Foo := foo : Type -> Foo.`, again this may be difficult to detect though.

**Comment by herbelin** [comment]
> Indeed this is a tricky case as thinking about it led me to #9294 so I'm not sure if we can detect it. Maybe if we do like with the "could not define as primitive record" warning where we look at what the kernel inferred.

It seems to me that we should first clarify what is the expected definition of template polymorphism. Is the discriminating feature of template polymorphism only that level parameters are instantiated by the default levels inferred from typing the parameters of the inductive type rather than using fresh levels together with appropriate constraints (and cumulativity to tell that the exact level of these fresh universes do not matter at the end)?

Another remark: it is surprising that we are allowed to define
```coq
Inductive Foo@{i} (A:Type@{i}) : Type := foo : A -> Foo A.
```
but then, `Check Foo@{Set}` fails with `Universe instance should have length 0`.

If the difference between template polymorphism vs full polymorphism is about automatic inference of "canonical" values for level parameters vs using fresh variables by default when no `@{}` is used, why to allow an explicit `@{}` only for full polymorphism?

**Comment by SkySkimmer** [comment]
>Is the discriminating feature of template polymorphism

There's no such thing, what there is is a discriminating implementation.

**Comment by herbelin** [comment]
> ... rather than using fresh levels together with appropriate constraints ...

Actually, minimization in "full" polymorphism is already minimizing the fresh levels as template polymorphism is doing. Let's for instance consider
```
Polymorphic Inductive Foo (A : Type) : Type :=  foo : A -> Foo A.
Set Printing Universes.
Check Foo nat.
(* Foo@{Set} nat : Set *)
```
where we can see that minimization has indeed played the rôle of template polymorphism (with the additional constraint to be at least >= `Set`).

I had a quick look to the implementation. Apparently, the first step of template polymorphism (i.e. the computation of a `max` of constraints over the levels of the conclusion of the parameters plus impact of isolated occurrences of `Type` in the constructors) is still in used even for full polymorphism (function `ComInductive.inductive_levels`).

It is the second step of template polymorphism (i.e. the instantiation of the levels of the conclusion by inferred levels) which is (more or less?) performed by minimization.

> There's no such thing, what there is is a discriminating implementation.

I don't know well the specification of the minimization algorithm but, would it then makes sense to replace the second step of template polymorphism (i.e. `type_of_inductive_knowing_parameters`) by a variant of minimization which accepts on demand to minimize a level without lower constraints to `Prop`.

The latter would fix #9294. (Alternatively, a solution to fix #9294 would be to take into account the graph of constraints in `type_of_inductive_knowing_parameters`).

Side question: How complicated would it be to display the graph of constraints of an inductive type when doing `Print` in `Set Printing Universes` mode, as we do for `Definition`?

Back to minimization, if `type_of_inductive_knowing_parameters` could be replaced by minimization, one would not need anymore to have a specific warning for implicit template polymorphism. The relevant warnings would be:
- a warning for implicit use of polymorphism (since there would be only one kind of polymorphism)
- a warning for allowing minimization to minimize to `Prop`
- a warning for restricting polymorphism to only parameters.

(Then, "template" would be a short-hand for standard polymorphism with polymorphism restricted to the parameters of the inductive type and minimization to `Prop` allowed.)

Again, I don't know well enough minimization to understand if the above can be made precise so that template can be subsumed by minimization, but assuming that minimization is made formal enough and that this becomes possible, what would be missing to reduce the concept of template polymorphism to the one of regular polymorphism and to get rid of `type_of_inductive_knowing_parameters`?

**Comment by SkySkimmer** [comment]
>Side question: How complicated would it be to display the graph of constraints of an inductive type when doing Print in Set Printing Universes mode, as we do for Definition?

Looks like some weird codepaths there, About prints constraints for any polymorphic thing whereas Print prints constraints for any definition regardless of polymorphism.

**Comment by SkySkimmer** [comment]
>what would be missing to reduce the concept of template polymorphism to the one of regular polymorphism

Instantiation by Prop mostly I think

**Comment by herbelin** [comment]
> Instantiation by Prop mostly I think

Not technically difficult then. Could we make a program of what the next steps are?

**Comment by JasonGross** [comment]
As a moderately heavy user of both universe polymorphism and template polymorphism (moreso in the past when I was doing category theory than now), I will chime in to say that in my mental model, the differences between universe polymorphism and template polymorphism are:
- template polymorphism is ~broken for partially applied inductives; template-polymorphic inductives should always be in eta-long form (whereas universe polymorphic inductives do not have this restriction)
- template polymorphism allows things like `prod True True : Prop`
- template polymorphism is only polymorphic over the parameters of an inductive type, not also the fields

(Question: Does `constr_eq` consider different universe-instantiations of polymorphic `prod` to be the same?  Different universe-instantiations of template-polymorphic `prod` to be the same?)

I am not currently aware of any developments that break when replacing template-polymorphic inductives with full universe-polymorphic inductives, unless they take advantage of `Prop`-instantiation.

So my current view is that we should definitely do your Option 1 for Minimization à la template to `Prop`, and that there should be a compatibility setting (even if only for the compat files) for automatically turning on polymorphism for inductives, even if it is off for everything else.  I am indifferent to whether or not there is an option for making an inductive polymorphic over only its parameters, and not its constructors.

**Comment by mattam82** [comment]
I think your option 1 could work, doing minimization to Prop on levels which have no explicit lower bound. That would mean that flexible universe variables would be allowed to be unified with Prop I suppose. Appart from being careful that we’re not breaking invariants in the uGraph about the lower bounds on universes and removing the restriction that instances cannot contain Prop I guess it’s not hard to implement. I also believe with this done you could see template as a restriction of full polymorphism to the apparent flexible universe variables in the parameters. One thing that might differ is how  `Ind (forall x : A, B x)` is treated, I.e, if the inferred level of the parameter is an algebraic, what happens?

**Comment by herbelin** [comment]
> I think your option 1 could work, doing minimization to Prop on levels which have no explicit lower bound. That would mean that flexible universe variables would be allowed to be unified with Prop I suppose. Appart from being careful that we’re not breaking invariants in the uGraph about the lower bounds on universes and removing the restriction that instances cannot contain Prop I guess it’s not hard to implement. I also believe with this done you could see template as a restriction of full polymorphism to the apparent flexible universe variables in the parameters.

Yes. (But to be clear, I'm just talking about an optional behavior of `Polymorphic`, not the default behavior.)

> One thing that might differ is how Ind (forall x : A, B x) is treated, I.e, if the inferred level of the parameter is an algebraic, what happens?

You mean, say, in:
```coq
Polymorphic Inductive list A := nil | cons : A -> list A -> list A.
Check fun A B => list (A -> B).
(*
fun (A : Type@{Top.16}) (B : Type@{Top.17}) =>
list@{Top.15} (A -> B)
     : Type@{Top.16} -> Type@{Top.17} -> Type@{max(Set,Top.15)}
   {Top.17 Top.16 Top.15} |= Top.16 <= Top.15
                             Top.17 <= Top.15 *)
```
I don't know well minimization, but could it be inferred that `Top.15` can be minimized to `max(Top.16,Top.17)`?

**Comment by SkySkimmer** [comment]
>I don't know well minimization, but could it be inferred that Top.15 can be minimized to max(Top.16,Top.17)?

It can't because it's used in `list@{Top.15}` and in that position algebraics aren't allowed.

**Comment by mattam82** [comment]
Indeed, and lifting that restriction would be quite challenging I think. But it might be very rarely used. In the really relevant case where the application of the inductive to the forall... parameter would fall down in Prop, the product itself would have type Prop.

**Comment by herbelin** [comment]
> It can't because it's used in list@{Top.15} and in that position algebraics aren't allowed.

I see. Then, if not allowed, I guess it is to prevent possible having to deal with equations of the form `u < max(v,w)`?

In template polymorphism, I'd have to check precisely but I don't see how lower constraints on the level of the conclusion of a parameter could arrive (if disallowing the kind of sharing that #9294 uses). Would it be possible to allow algebraics in situations where template polymorphism was allowing them? Or would it break somewhere else?
